support/k8s/spire-adm-webhook/webhook.go

```diff
@@ -0,0 +1,225 @@
+// configuration has filters in place that prevent the webhook from being
+// called for requests other than these but just in case, we filter here as
+// well.
+func (w *Webhook) reviewAdmission(ctx context.Context, req *admv1beta1.AdmissionRequest) error {
+	switch req.Namespace {
+	case metav1.NamespacePublic, metav1.NamespaceSystem:
+		return nil
+	}
+
+	if req.Kind != (metav1.GroupVersionKind{Version: "v1", Kind: "Pod"}) {
+		return nil
+	}
+
+	switch req.Operation {
+	case admv1beta1.Create:
+		pod := new(corev1.Pod)
+		if err := json.Unmarshal(req.Object.Raw, pod); err != nil {
+			return errs.New("unable to unmarshal %s/%s object: %v", req.Kind.Version, req.Kind.Kind, err)
+		}
+		return w.createPodEntry(ctx, pod)
+	case admv1beta1.Delete:
+		return w.deletePodEntry(ctx, req.Namespace, req.Name)
+	}
+
+	return nil
+}
+
+func (w *Webhook) createPodEntry(ctx context.Context, pod *corev1.Pod) error {
+	if w.c.PodLabel != "" {
+		// the webhook has been been configured with a pod label. if the pod
+		// has that label, use the value to construct the pod entry. otherwise
+		// ignore the pod altogether.
+		if labelValue, ok := pod.Labels[w.c.PodLabel]; ok {
+			return w.createPodEntryByLabel(ctx, pod, w.c.PodLabel, labelValue)
+		}
+		return nil
+	}
+
+	// the webhook has not been configured with a pod label. create an entry
+	// based on the service account.
+	return w.createPodEntryByServiceAccount(ctx, pod)
+}
+
+func (w *Webhook) createPodEntryByLabel(ctx context.Context, pod *corev1.Pod, labelKey, labelValue string) error {
+	return w.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: w.nodeID(),
+		SpiffeId: w.makeID("%s", labelValue),
+		Selectors: []*common.Selector{
+			namespaceSelector(pod.Namespace),
+			podNameSelector(pod.Name),
+		},
+	})
+}
+
+func (w *Webhook) createPodEntryByServiceAccount(ctx context.Context, pod *corev1.Pod) error {
+	return w.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: w.nodeID(),
+		SpiffeId: w.makeID("ns/%s/sa/%s", pod.Namespace, pod.Spec.ServiceAccountName),
+		Selectors: []*common.Selector{
+			namespaceSelector(pod.Namespace),
+			podNameSelector(pod.Name),
+		},
+	})
+}
+
+func (w *Webhook) deletePodEntry(ctx context.Context, namespace, name string) error {
+	log := w.c.Log.WithFields(logrus.Fields{
+		"ns":  namespace,
+		"pod": name,
+	})
+
+	entries, err := w.c.R.ListBySelectors(ctx, &common.Selectors{
+		Entries: []*common.Selector{
+			namespaceSelector(namespace),
+			podNameSelector(name),
+		},
+	})
+	if err != nil {
+		return errs.New("unable to list by pod entries: %v", err)
+	}
+
+	log.WithField("count", len(entries.Entries)).Debug("Deleting entries")
+
+	var errGroup errs.Group
+	for _, entry := range entries.Entries {
+		_, err := w.c.R.DeleteEntry(ctx, &registration.RegistrationEntryID{
+			Id: entry.EntryId,
+		})
+		if err != nil {
+			log.WithField("err", err.Error()).Debug("Failed deleting pod entry")
```

Comment:
This should probably be logged at error level

Reply:
probably `log.WithError(err)` too?

Reply:
👍